optuna/_imports.py

```diff
@@ -0,0 +1,97 @@
+from types import TracebackType
+from typing import Optional
+from typing import Tuple
+from typing import Type
+from typing import Union
+
+
+class _DeferredExceptionContextManager(object):
+    """Context manager to defer exceptions.
+
+    Args:
+        catch:
+            Exception types to defer.
+        message:
+            Message to prefix in deferred exception.
+
+    """
+
+    def __init__(self, catch: Union[Tuple[()], Tuple[Type[Exception], ...]], message: str) -> None:
+        self._catch = catch
+        self._message = message
+        self._caught_exception = None  # type: Optional[Tuple[Type[Exception], Exception]]
+
+    def __enter__(self) -> "_DeferredExceptionContextManager":
+        """Enter the context manager.
+
+        Returns:
+            Itself.
+
+        """
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[Exception]],
+        exc_value: Optional[Exception],
+        traceback: Optional[TracebackType],
```

Comment:
Currently, the `traceback` argument is not used. This is not a strong opinion, but how about showing it with the error message?
IMO, it helps users to find the issue of the code.
For example, the current message shown in the console is as follows:

```console
>>> optuna.integration.ChainerPruningExtension(None, None, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/optuna/optuna/integration/chainer.py", line 53, in __init__
    _imports.check()
  File "/tmp/optuna/optuna/_imports.py", line 88, in check
    raise exc_type(msg)
ImportError: Failed to import an optional package. No module named 'chainer'.
```

It does not contain the line where `chainer` was imported, and users need to guess it.
On the other hand, `traceback` has the line as follows:
```console
  File "/tmp/optuna/optuna/integration/chainer.py", line 5, in <module>
    import chainer
```

Reply:
Thanks, that makes sense. We can probably nest the exceptions using `raise ... from ...`. I'll push the change. I think it'll include enough information to satisfy https://github.com/optuna/optuna/pull/1315#pullrequestreview-423254751 although I agree that the former was friendlier.

Reply:
Example after https://github.com/optuna/optuna/pull/1315/commits/43a31f67e6af930f94abe9744e4e5a52c2c1ff86

```
Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 5, in <module>
    import chainer
ModuleNotFoundError: No module named 'chainer'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/study.py", line 732, in _run_trial
    result = func(trial)
  File "test_try_import.py", line 6, in objective
    ChainerPruningExtension(trial, "validation/main/accuracy", (1, "epoch"))
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 53, in __init__
    _imports.check()
  File "/Users/hvy/repos/optuna/optuna/_imports.py", line 80, in check
    raise exc_type(self._message) from exc_value
ModuleNotFoundError: Failed to import an optional package. See causing exceptions for details.
Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 5, in <module>
    import chainer
ModuleNotFoundError: No module named 'chainer'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_try_import.py", line 12, in <module>
    study.optimize(objective, n_trials=10)
  File "/Users/hvy/repos/optuna/optuna/study.py", line 334, in optimize
    func, n_trials, timeout, catch, callbacks, gc_after_trial, None
  File "/Users/hvy/repos/optuna/optuna/study.py", line 677, in _optimize_sequential
    self._run_trial_and_callbacks(func, catch, callbacks, gc_after_trial)
  File "/Users/hvy/repos/optuna/optuna/study.py", line 708, in _run_trial_and_callbacks
    trial = self._run_trial(func, catch, gc_after_trial)
  File "/Users/hvy/repos/optuna/optuna/study.py", line 732, in _run_trial
    result = func(trial)
  File "test_try_import.py", line 6, in objective
    ChainerPruningExtension(trial, "validation/main/accuracy", (1, "epoch"))
  File "/Users/hvy/repos/optuna/optuna/integration/chainer.py", line 53, in __init__
    _imports.check()
  File "/Users/hvy/repos/optuna/optuna/_imports.py", line 80, in check
    raise exc_type(self._message) from exc_value
ModuleNotFoundError: Failed to import an optional package. See causing exceptions for details.
```

Reply:
Thanks for the update. My concerns have been addressed by this fix.

Reply:
Awesome! Thank you for your update.